domain/resource/state/resource.go

```diff
@@ -407,173 +419,568 @@ WHERE uuid = $resourceIdentity.uuid`,
 		return resource.Resource{}, errors.Capture(err)
 	}
 
-	return resourceOutput.toResource(), nil
+	return resourceOutput.toResource()
 }
 
-// SetResource adds the resource to blob storage and updates the metadata.
-func (st *State) SetResource(
+// StoreResource records a stored resource along with who retrieved it.
+// Returns [resourceerrors.ResourceNotFound] if the resource UUID cannot be found.
+// Returns [resourceerrors.StoredResourceNotFound] if the stored resource at the
+// storageID cannot be found.
+// Returns [resourceerrors.ResourceAlreadyStored] if the resource is already
+// associated with a stored resource blob.
+// Returns [resourceerrors.RetrievedByTypeNotValid] if the retrieved by type is
+// invalid.
+func (st *State) StoreResource(
 	ctx context.Context,
-	config resource.SetResourceArgs,
-) (resource.Resource, error) {
-	return resource.Resource{}, nil
+	resourceUUID coreresource.UUID,
+	storageID coreresourcestore.ID,
+	retrievedBy string,
+	retrievedByType resource.RetrievedByType,
+	incrementCharmModifiedVersion bool,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		kind, err := st.getResourceType(ctx, tx, resourceUUID)
+		if err != nil {
+			return errors.Errorf("getting resource type: %w", err)
+		}
+
+		switch kind {
+		case charmresource.TypeFile:
+			err = st.insertFileResource(ctx, tx, resourceUUID, storageID)
+			if err != nil {
+				return errors.Errorf("inserting stored file resource information: %w", err)
+			}
+		case charmresource.TypeContainerImage:
+			err = st.insertImageResource(ctx, tx, resourceUUID, storageID)
+			if err != nil {
+				return errors.Errorf("inserting stored container image resource information: %w", err)
+			}
+		default:
+			return errors.Errorf("unknown resource type: %q", kind.String())
+		}
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		if retrievedBy != "" {
+			err := st.insertRetrievedBy(ctx, tx, resourceUUID, retrievedBy, retrievedByType)
+			if err != nil {
+				return errors.Errorf("inserting retrieval information for resource %s: %w", resourceUUID, err)
+			}
+		}
+
+		if incrementCharmModifiedVersion {
+			err := st.incrementCharmModifiedVersion(ctx, tx, resourceUUID)
+			if err != nil {
+				return errors.Errorf("inserting retrieval information for resource %s: %w", resourceUUID, err)
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
```

Comment:
**todo:** I am not sure the capture give more context. returning the error directly seems good enough.